policy/options.go

```diff
@@ -0,0 +1,460 @@
+		}
+		e.permittedURIDomains = normalizedURIDomains
+		return nil
+	}
+}
+
+func WithPermittedURIDomain(domain string) NamePolicyOption {
+	return func(e *NamePolicyEngine) error {
+		normalizedURIDomain, err := normalizeAndValidateURIDomainConstraint(domain)
+		if err != nil {
+			return fmt.Errorf("cannot parse permitted URI domain constraint %q: %w", domain, err)
+		}
+		e.permittedURIDomains = []string{normalizedURIDomain}
+		return nil
+	}
+}
+
+func WithExcludedURIDomains(domains []string) NamePolicyOption {
+	return func(e *NamePolicyEngine) error {
+		normalizedURIDomains := make([]string, len(domains))
+		for i, domain := range domains {
+			normalizedURIDomain, err := normalizeAndValidateURIDomainConstraint(domain)
+			if err != nil {
+				return fmt.Errorf("cannot parse excluded URI domain constraint %q: %w", domain, err)
+			}
+			normalizedURIDomains[i] = normalizedURIDomain
+		}
+		e.excludedURIDomains = normalizedURIDomains
+		return nil
+	}
+}
+
+func WithExcludedURIDomain(domain string) NamePolicyOption {
+	return func(e *NamePolicyEngine) error {
+		normalizedURIDomain, err := normalizeAndValidateURIDomainConstraint(domain)
+		if err != nil {
+			return fmt.Errorf("cannot parse excluded URI domain constraint %q: %w", domain, err)
+		}
+		e.excludedURIDomains = []string{normalizedURIDomain}
+		return nil
+	}
+}
+
+func WithPermittedPrincipals(principals []string) NamePolicyOption {
+	return func(g *NamePolicyEngine) error {
+		// TODO(hs): normalize and parse principal into the right type? Seems the safe thing to do.
+		g.permittedPrincipals = principals
+		return nil
+	}
+}
+
+func WithExcludedPrincipals(principals []string) NamePolicyOption {
+	return func(g *NamePolicyEngine) error {
+		// TODO(hs): normalize and parse principal into the right type? Seems the safe thing to do.
+		g.excludedPrincipals = principals
+		return nil
+	}
+}
+
+func networkFor(ip net.IP) *net.IPNet {
+	var mask net.IPMask
+	if !isIPv4(ip) {
+		mask = net.CIDRMask(128, 128)
+	} else {
+		mask = net.CIDRMask(32, 32)
+	}
+	nw := &net.IPNet{
+		IP:   ip,
+		Mask: mask,
+	}
+	return nw
+}
+
+func isIPv4(ip net.IP) bool {
+	return ip.To4() != nil
+}
+
+func normalizeAndValidateDNSDomainConstraint(constraint string) (string, error) {
+	normalizedConstraint := strings.ToLower(strings.TrimSpace(constraint))
+	if normalizedConstraint == "" {
+		return "", fmt.Errorf("contraint %q can not be empty or white space string", constraint)
+	}
+	if strings.Contains(normalizedConstraint, "..") {
+		return "", fmt.Errorf("domain constraint %q cannot have empty labels", constraint)
+	}
+	if strings.HasPrefix(normalizedConstraint, ".") {
+		return "", fmt.Errorf("domain constraint %q with wildcard should start with *", constraint)
+	}
+	if strings.LastIndex(normalizedConstraint, "*") > 0 {
+		return "", fmt.Errorf("domain constraint %q can only have wildcard as starting character", constraint)
+	}
+	if normalizedConstraint[0] == '*' && normalizedConstraint[1] != '.' {
```

Comment:
Panic if `normalizedConstraint` length is 1.

Reply:
I've made it a bit more safe in a2cfbe3d546404932ca15bd8b7db981932a9a179.